### ISSUE TITLE
[HOLD for payment 2023-04-27] [$2000] Split -  'Split bill' incorrectly divides bills for some currencies

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#15878]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open app and log in
2. Click the FAB button and select Split Bill
3. Select currency with a low exchange rate (for example, AFN or JPY)
4. Enter an amount such that when divided by the number of attendees, fractional values are obtained. For example: 103 for 3 attendees, or 10 JPY between 4 attendees.
5. Select attendees
6. Check the amount for each attendee and sum these numbers

## Expected Result:

- The amount obtained by summing up the numbers for each participant must be equal to the original amount that was entered in the BNP, even if this results in slightly different amounts for the different participants.
- We shouldn't send "cents" to the backend if the unit doesn't show "cents" in the UI. For example, if we split JPY, we should make sure we divide the amounts in integers.

10 JPY between 4 attendees should be split like 3, 3, 3 and 1 (remainder: 10 - 3 - 3 - 3), and the request to the backend should have the amounts 300, 300, 200 and 200. Each part that is not the remainder is just the `Math.round(total / parts)`.

## Actual Result:
The sum of divided bills does not equal the original number. Looks like the app divides with rounding and then discards all decimal places for currencies with low exchange rate like JPY.
In our case we get 34x3 = 102 and not 103 as it should.

The request `SplitBillAndOpenReport` has decimal amounts in it:

<img width="486" alt="image" src="https://user-images.githubusercontent.com/87341702/228504387-973898b5-690d-4ff4-80ba-380e44cbf958.png">

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.2.82.3

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

<img src="https://user-images.githubusercontent.com/93399543/224502942-20fcddc8-3e32-426f-9055-1b666d66c013.mp4" width="40%" height="40%">

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~013aa2514a7501c2d0</li>
        <li>Upwork Job ID: 1635977229585281024</li>
        <li>Last Price Increase: 2023-03-29</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @CortneyOfstad (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by CortneyOfstad** [comment]
Diving into this today!

**Comment by CortneyOfstad** [comment]
So when I test this, I am able to get the currency to add up appropriately with ANG, but not JPY. With ANG, I requested $103 from three others, breaking the total amount down to $25.75 each, which was correct. 

However, when I utilize JPY, it changes the amount to 26 yen each, which adds up to 104, not 103. Going to get eyes on this. 

It's odd that only some currencies are affected.

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~013aa2514a7501c2d0

**Comment by MelvinBot** [comment]
Current assignee @CortneyOfstad is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @aldo-expensify (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by tienifr** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The split bill is splitting wrong amount for some currencies like JPY.

### What is the root cause of that problem?
Our function https://github.com/Expensify/App/blob/4986dea08e3992319dd7581b637ad531ebc5dd15/src/libs/IOUUtils.js#L10 to split amount is incorrect, it's multiplying 100 for all currencies, regardless of its exponent.

For some currencies like JPY with exponent 0, `25.75 JPY` does not mean anything because there's no such thing as a cent in JPY.
USD has exponent 2, so `25.75 USD` makes sense.

When we display the value, we're using the standard `Intl.NumberFormat` function so it's displaying  `25.75 JPY` correctly as `26 JPY`

### What changes do you think we should make in order to solve the problem?

We should add a `currency` params to `calculateAmount` and multiply the correct exponent.

We can update https://github.com/Expensify/App/blob/4986dea08e3992319dd7581b637ad531ebc5dd15/src/libs/IOUUtils.js#L16 to ` const iouAmount = Math.round(parseFloat(total * (10 ** currencyDigitsData[currency])));`

The currencyDigitsData can be taken from here https://github.com/formatjs/formatjs/blob/62ec0b3338dacc54340c559a11128a50e02d95a2/packages/intl-numberformat/src/currency-digits.generated.ts.

### What alternative solutions did you explore? (Optional)
NA

**Comment by allroundexperts** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The split bill option is not correctly splitting bills for some currencies.

### What is the root cause of that problem?
As correctly pointed out by @tienifr, the issue arises from the fact that currencies like JPY don't have a minor unit.

### What changes do you think we should make in order to solve the problem?

For the currencies that do not have a minor unit, we can keep on using `Intl.NumberFormat` as its not rounding incorrectly. However, we can check the result produced from `Intl.NumberFormat` adds up to the entered amount. If it does add up, we do nothing. If it does not add up (due to the number being rounded), we can simply remove the extra value from the either of the participants and show users a note that why this is necessary. 

I think even if we use libraries such as `https://dinerojs.com/`, it will just round it to the lower side and the numbers won't really add up to the total (it would be 1 less than the original amount if rounded on the lower side).

### What alternative solutions did you explore? (Optional)
N/A

**Comment by hoangzinh** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Split - 'Split bill' incorrectly divides bills for some currencies like JPY

### What is the root cause of that problem?
As other people explained, we're using `Intl.NumberFormat` to format currency, unfortunately, some currencies doesn't have fraction digit. So it automatically round the number by default.


### What changes do you think we should make in order to solve the problem?
I think we can add an option that is supported by `Intl.NumberFormat`
```diff
this.props.numberFormat(IOUUtils.calculateAmount(participants, this.props.iouAmount) / 100, {
    style: 'currency',
    currency: this.props.iou.selectedCurrencyCode,
+   minimumFractionDigits: 2,
}),
```

That means, we alway have at least 2 fraction digits when displaying (even with 00)
![Screenshot 2023-03-16 at 00 48 46](https://user-images.githubusercontent.com/9639873/225398456-708e3c6d-c416-4e11-8f78-54a3fde099b8.png)
![Screenshot 2023-03-16 at 00 49 00](https://user-images.githubusercontent.com/9639873/225398462-1e96e11e-3126-4dfc-a49b-3fd62966c79d.png)


### What alternative solutions did you explore? (Optional)
Only set `minimumFractionDigits` when formatting if the amount is a decimal type.

**Comment by CortneyOfstad** [comment]
@aldo-expensify How are we feeling about the above proposal?

**Comment by aldo-expensify** [comment]
I think this may not be an issue really. As @aimane-chnaif is saying above, if the currency doesn't have "cents", it doesn't make sense to include decimals.

I'm asking what is the expected rounding in this cases (always up, closest integer).

**Comment by aldo-expensify** [comment]
Yes, this is expected behaviour for currencies without "cents".

I'm closing this.

**Comment by aldo-expensify** [comment]
> @aldo-expensify
> 
> > if the currency doesn't have "cents", it doesn't make sense to include decimals.
> 
> The currencies without "cents" are currently sent to the back-end **with** "cents", in the above case it shows as 34, 34, 34 in the UI but it will send 3434, 3433 3433 cents to the API, so it's incorrect.


Ahh, good point. I'll open this again then, seems like we need to look into it further.

Thanks!

**Comment by hayata-suenaga** [comment]
Japanese currency actually has a unit called ["銭" (sen)](https://www.japan-guide.com/e/e2196.html#:~:text=One%20yen%20corresponds%20to%20100,yen%20and%20500%20yen%20denominations.) which is used for very precious financial transactions (ex. for currency exchange).

Although we don't use it for everyday purpose, it's also not weird to see decimal numbers with ¥. IMO, the UI should reflect the real price that is requested on API.

If they have bank or card statements, the decimal might appear. It's better to make the price displayed on the App and bank/card statements to match exactly.

**Comment by tienifr** [comment]
@hayata-suenaga As per my understanding, `splitting bills` here means the actual money will be transfered between the people to pay for the bill. Also it seems not possible to do a bank transfer of decimal JPY (like 1.5), I tried with the Wise app, it must always be a non-decimal number of JPY. 
> it seems not possible to do a bank transfer of decimal JPY

If this is indeed the case, I don't think we should allow decimal JPY since people cannot really use it in the context of using our app.

**Comment by MelvinBot** [comment]
Upwork job price has been updated to $1000

**Comment by aldo-expensify** [comment]
> Another option would be to split "not equal" similar to dividing cents
i.e. $5.00 = 167 + 166 + 166 cents, 500 JPY = 167 + 166 + 166 JPY

yes, I think that will be the way to go for the first problem

**Comment by hayata-suenaga** [comment]
> Maybe you can't have (sen) in your bank account, can you?

I'm actually not sure. I haven't touched my jp bank account for years...

**Comment by aldo-expensify** [comment]
Ok, so lets forget for now about the JPY having decimals or not, we can focus on working on what this issue was initially about. If we think the JPY should not have decimals, that is a different issue and should be reported separately.

@aimane-chnaif you can review proposals, the only problem to solve is the sum not adding up correctly.

I'll mark as off topic all comments related to JPY having decimal places.

**Comment by aldo-expensify** [comment]
Hmm argh, disregard my comment above, this is kind of dependent on the decimal places problem.

**Comment by aldo-expensify** [comment]
I know that we can do cents when we split the amount equally, but I'm thinking that we shouldn't split in cents if the currency doesn't show cents in the UI. For example, for JPY, if we split 4 JPY between 3 people, the splits should be 1, 1 and 2, and not 1.33, 1.33 and 1.34.

Do you agree @aimane-chnaif @hayata-suenaga ?

**Comment by hayata-suenaga** [comment]
@aldo-expensify I did some digging into what existing apps in Japan handle this situation, but I couldn't find any.

I believe your solution is the most feasible one to me.

**Comment by aldo-expensify** [comment]
> @aimane-chnaif are we opening for proposals to implement what we are expecting above ^?

Yes, I updated the expected results.

**Comment by tienifr** [comment]
Thanks @aldo-expensify!

My earlier proposal [here](https://github.com/Expensify/App/issues/15878#issuecomment-1469972557) will do exactly what we want.

**Comment by aldo-expensify** [comment]
@aimane-chnaif friendly bump for reviewing proposals

**Comment by tienifr** [comment]
> Also it seems we're sending things like 2600 JPY "cents" to the BE, if we don't want to fix the BE (which I think we should do) then in addition to the above change we need to convert 26 JPY from calculateAmount to 2600 JPY "cents" (by dividing by getCurrencyMinorUnits(currency) and multiply with 100) before sending to the BE.

@aldo-expensify yup for that we can apply this part of my proposal.

**Comment by tienifr** [comment]
> I completed an example with JPY: 10 JPY between 4 attendees should be split like 3, 3, 2 and 2, and the request to the backend should have the amounts 300, 300, 200 and 200.

@aldo-expensify it looks like you want to change how we split the amount as well rather than just dealing with the special currencies. 
For example for regular currencies like `USD`, if we split 0.1 USD (10 cents) to 4 people, it will be split to 0.01, 0.03, 0.03, 0.03 (current staging/prod is like this, the user who splits will get the 0.01 and the rest will have the same split). While you're expecting it will be split to 0.03, 0.03, 0.02, 0.02.

This seems to be different from the original issue here which only relates to some special currencies and whether they have cents or not. Can you confirm we want to do this (change how we split even for regular currencies like USD)?

**Comment by aldo-expensify** [comment]
> > I completed an example with JPY: 10 JPY between 4 attendees should be split like 3, 3, 2 and 2, and the request to the backend should have the amounts 300, 300, 200 and 200.
> 
> @aldo-expensify it looks like you want to change how we split the amount as well rather than just dealing with the special currencies. For example for regular currencies like `USD`, if we split 0.1 USD (10 cents) to 4 people, it will be split to 0.01, 0.03, 0.03, 0.03 (current staging/prod is like this, the user who splits will get the 0.01 and the rest will have the same split). While you're expecting it will be split to 0.03, 0.03, 0.02, 0.02.
> 
> This is different from the original issue here which only relates to some special currencies and whether they have cents or not. Can you confirm we want to do this (change how we split even for regular currencies like USD) as part of this ticket or create another bug for this?

Sorry, that was my mistake, it is not the intention to change how it works for USD. I'll update my example to follow the same pattern. Splitting 10 JPY between 4 people should be 3, 3, 3 and 1 if that is more consistent with the current split in USD.

**Comment by tienifr** [comment]
Thanks for confirming @aldo-expensify, I'll report the split logic issue as a separate bug. I think it's not right since if we split (0.06 to 4 people, the split will be [0, 0.02, 0.02, 0.02]) which basically means "you guys take care of the bill, I won't pay anything" 😄.

**Comment by aldo-expensify** [comment]
> Thanks for confirming @aldo-expensify, I'll report the split logic issue as a separate bug. I think it's not right since if we split 0.06 to 4 people, the split will be [0, 0.02, 0.02, 0.02] which basically means "you guys take care of the bill, I won't pay anything" 😄.

haha, I had the same thought, but I decided to ignore it because this is always about tiny amounts I think. You can still reported, maybe other people think this should be changed. If you report it, please link this issue as related.

**Comment by tienifr** [comment]
# Proposal

[Proposal updated](https://github.com/Expensify/App/issues/15878#issuecomment-1469972557) for clarify in implementation.

**Comment by aldo-expensify** [comment]
> > 10 JPY between 4 attendees should be split like 3, 3, 3 and 1 (remainder: 10 - 3 - 3 - 3)
> 
> @aldo-expensify Sorry, but it sounds incorrect right?

it may be, but my intention is to keep the scope of this issue small and that current split logic can be discussed/fixed in an another issue. I had a small chat with @JmillsExpensify about this and splitting these tiny amounts into many participants is just an edge case we may not care about.

Here we are just dealing with the cents appearing in currencies that don't support it.

**Comment by JmillsExpensify** [comment]
Yeah agree. I'm not really clear how/whey this case of splitting 10 JPY would ever exist in the real world.

**Comment by ShogunFire** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

The split doesn't work well when the currencies have a low exponent 

### What is the root cause of that problem?

The method IOUUtils.calculateAmount doesn't take into account the currency exponent so it split the amount in a way that is not possible in reality

### What changes do you think we should make in order to solve the problem?

CalculateAmount should take a new argument currency. Then here is the algorithm that I propose:


Let's say the amount is 103 that has to be split among 3 persons.

- We divide 103 by 3 which gives around 34,33

- If currency exponent = 0 we remove everything after the "," so 34 we add that to the amount each participant has to pay.


- We multiply this number by the number of participants.
34 * 3 = 102 this is the amount that has already been split.

- We substract the original amount by this amount to get the amount that still need to be split:
103-102 = 1 this number is always inferior to the number of participants.


- We loop through the participants and give them 1 more to pay until there is nothing left to split.

If for example currency exponent is 0,1 it's the same algorithm but we remove everything after the first decimal instead of after "," and at the end we add 0,1 instead of 1. 


### What alternative solutions did you explore? (Optional)

**Comment by aldo-expensify** [comment]
> @tienifr thanks for following discussions which made scope of this issue more clear.
> 
> Some feedback from your [proposal](https://github.com/Expensify/App/issues/15878#issuecomment-1469972557):
> 
> > Get the currency minor units data. It can be taken from the intl-numberformat that we're using here https://github.com/formatjs/formatjs/blob/62ec0b3338dacc54340c559a11128a50e02d95a2/packages/intl-numberformat/src/currency-digits.generated.ts.
> 
> I am not sure if we should use data from this library. i.e. even main currencies (USD, EUR, GBP) don't exist. And we can't confirm that missing currencies always have 2 decimal digits.
> 
> > We can use another library that deals specifically with number like https://dinerojs.com/ to split the bill/, then we'll have the correct currency handling out of the box.
> 
> Using another 3rd party is the last option. Let's try to find solution with currently imported libraries as much as possible.

Agreed, I think we currently use [Intl.NumberFormat](https://github.com/Expensify/App/blob/6151893d35ea4170028b96fd3ef536ba8a81d945/src/libs/NumberFormatUtils.js#L1-L3) to format currency? it would be ideal to use the same and not a 3rd party library so we don't get inconsistencies where`Intl` doesn't use decimals from some currency and the 3rd party library does.

**Comment by tienifr** [comment]
> I am not sure if we should use data from this library. i.e. even main currencies (USD, EUR, GBP) don't exist. And we can't confirm that missing currencies always have 2 decimal digits.

@aimane-chnaif you can see from [this line](https://github.com/formatjs/formatjs/blob/9c4eca1513b755bf41d703b215a7293420185fdc/packages/ecma402-abstract/src/NumberFormat/CurrencyDigits.ts#L10) to see how the `Intl.NumberFormat` that we're using get the digits, first it tries to get from the [list here](https://github.com/formatjs/formatjs/blob/62ec0b3338dacc54340c559a11128a50e02d95a2/packages/intl-numberformat/src/currency-digits.generated.ts). If it doesn't appear there then it will default to 2.

I believe the library creator already makes sure that `missing currencies from the list always have 2 decimal digits`.

If you still have doubts, we can construct the list ourselves using the ISO Currency table here https://en.wikipedia.org/wiki/ISO_4217 to be 100% sure.

**Comment by tienifr** [comment]
@aimane-chnaif I think @aldo-expensify already mentioned that there should be no back-end change [here](https://github.com/Expensify/App/issues/15878#issuecomment-1488353542), just in case you missed it.

**Comment by ShogunFire** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/15878#issuecomment-1488782378)

**Comment by tienifr** [comment]
> currency: BHD
split 10, 0.05 to 3 people (including current user)
Expected result: (3.334, 3.333, 3.333), (0.016, 0.017, 0.017)

@aimane-chnaif here's the demo video you requested for this case

https://user-images.githubusercontent.com/113963320/228623640-c9b846f0-b4a7-4e94-ac98-947743cc141d.mp4

**Comment by aldo-expensify** [comment]
> > your solution splits this to (-1, 1, 1, 1)
> > I understand that we agreed division logic out of scope but at least avoid negative value
> 
> @aimane-chnaif Currently on staging splitting $0.02 to 4 people also give (-0.01, 0.01, 0.01, 0.01). I already reported the splitting logic issue [here](https://expensify.slack.com/archives/C049HHMV9SM/p1680092552607329). The solution will be to amend that logic as mentioned [here](https://github.com/Expensify/App/issues/15878#issuecomment-1488520625).
> 
> Are you ok to solve it in that reported issue rather than here?

ugh, I guess negative values are really not something we want, I'm not sure what kinds of problems that can bring either.

I think that fixing the splitting in a way that doesn't cause negatives and is more fair shouldn't be hard, but this may open another conversation too 😬 : what if the amount is so small that assigns 0 to some participants? (i.e. split 1 between 2 people)

@luacmartins can I have some thoughts about this since you worked in the splits :) ? What would happen if some parts end up being 0? do we care?

I'm still thinking that we shouldn't touch the splitting calculation here because it may domino effect into changing many things and this is an extreme edge case.

**Comment by luacmartins** [comment]
TBH I'm not sure what happens if you split $0 with others, I don't recall testing that during implementation. I imagine that it'd be ok and a way for people to keep track of requests, even if they are small. That being said, this seems like a pretty rare edge case, since I don't think most users would even bother splitting something that small. I'd say we don't need to worry about that now.

**Comment by luacmartins** [comment]
IMO all those split options are incorrect and I'd say we should do nothing in this case, given how rare this is. This seems like an issue that could be solved by uneven splits or a minimum amount based on the currency smallest denomination and number of participants.

**Comment by CortneyOfstad** [comment]
@aldo-expensify @aimane-chnaif Any updates here? 

Reassigning this to Bug Zero as I am OoO all next week 👍

**Comment by MelvinBot** [comment]
Triggered auto assignment to @kevinksullivan (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by tienifr** [comment]
@aldo-expensify Awaiting your approval 😎

**Comment by aldo-expensify** [comment]
Sorry, I missed it.

I looked into the proposals, and I agree with going with @tienifr 's [proposal](https://github.com/Expensify/App/issues/15878#issuecomment-1469972557)

About using https://github.com/formatjs/formatjs/blob/62ec0b3338dacc54340c559a11128a50e02d95a2/packages/intl-numberformat/src/currency-digits.generated.ts, I wonder if it is unecessary to have this hardcoded data and if we can just calculate the number of decimals on the fly using `Intl.formatNumber`. Why? we don't add unnecessary bytes to our package and we don't have to care about keeping the decimals up to date.

**Comment by MelvinBot** [comment]
📣 @tienifr You have been assigned to this job by @aldo-expensify!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by tienifr** [comment]
@aldo-expensify I don’t think there’s any straight-forward way to get the decimals of the currency using that ‘Intl.formatNumber’. There could be workarounds like try to format a dummy currency number (1.11111 for example) to try to get the decimal but I don’t think we want such workaround in our code. 

Also trying to deal with JS decimals is error-prone which is why we need to multiply by the cents to get the non-decimal amount.

cc @aimane-chnaif

**Comment by aldo-expensify** [comment]
It is true that `Intl.formatNumber` doesn't have a function to get this, but maybe something like this could be simple enough:

```js
function getDecimals(currencySymbol) {
    const formatted = Intl.NumberFormat('en', {style: 'currency', currency: currencySymbol}).format(0);
    const decimalPointIndex = formatted.indexOf('.');
    return decimalPointIndex === -1
        ? 0 // No decimals
        : formatted.length - decimalPointIndex - 1;
}
```

> Also dealing with JS decimals is error-prone which is why we need to multiply by the cents to get the non-decimal amount and operate from there.

I didn't understand this part, the calculations are happening in integer amounts anyway, right?

This made me notice that we don't really support more than 2 decimal places in the back end, and there are currencies that have more than 2. If a currency supports 3 decimals, we will have to round things to 2 decimals, not 3.

**Comment by aldo-expensify** [comment]
> does this mean that we should treat all >2 decimals as 2? i.e.
> 
> ```
>     return decimalPointIndex === -1
>         ? 0 // No decimals
>         : Math.min(formatted.length - decimalPointIndex - 1, 2);
> ```

Yes, I think that would work.

About the backend, I'm pretty sure that we work with numbers as integers, and the amounts are in cents. So 1 YEN in the backend is really stored as 100.

If a currency had 3 decimal places, I think we won't be able to store the last decimal, so we will have to round to the second. We can store 0.01 CLF as 1 in the backend, but we can't store 0.001 or 0.0001.

**Comment by tienifr** [comment]
@aimane-chnaif @aldo-expensify I really appreciate your contributions. I'll go with that.

**Comment by tienifr** [comment]
> It is true that Intl.formatNumber doesn't have a function to get this, but maybe something like this could be simple enough

@aldo-expensify @aimane-chnaif  I still feel that adding the decimal in [currencyList.json](https://github.com/Expensify/App/blob/4efde840cc7456d14a2f79185e95cd9d95f79b90/tests/unit/currencyList.json#L21) is better, we're only adding 1 more field for the currency there so the overhead in code is low (I think it will even add less bytes to the codebase than adding that manipulation function and all the unit tests for the currencies to make sure it's working correctly). Also that way is more performant than string manipulation using the `Intl.formatNumber`.

With that, if we still feel strongly about using the `Intl.formatNumber` then I'm happy to change it in the PR.

**Comment by aldo-expensify** [comment]
@tienifr I think it is fine to add it in the [currencyList.json](https://github.com/Expensify/App/blob/4efde840cc7456d14a2f79185e95cd9d95f79b90/tests/unit/currencyList.json#L21) as a new key. When I made that comment, I thought we were adding a completely new currency list, but I'm fine with reusing that.

**Comment by aldo-expensify** [comment]
@aimane-chnaif ah! good call, then I'm again thinking that we should stick with the function.

**Comment by aldo-expensify** [comment]
hum, now I'm wondering if the backend should be sending the number of decimals since it is sending other currency data.

**Comment by aldo-expensify** [comment]
> The concern is that what resource backend will be using to be consistent with frontend Intl.NumberFormat

The library in the backend is pretty consistent with `Intl.NumberFormat`, it only differs in some currency symbols, but the decimals are the same I think.

**Comment by aldo-expensify** [comment]
The currency list we send from the backend already contains a property call "decimals" for some of them. This property usually is `0` or missing. If it is missing, we are supposed to assume that there are 2 decimals.

This list is a bit outdated and there is a good number of currencies with 0 decimals that are missing `decimals: 0`. I'm updating this in the backend. After this is updated, @tienifr you can use the currency list from onyx to figure out the number of decimals.

**Comment by aldo-expensify** [comment]
> @aldo-expensify should we hold this until backend update deploy to staging?

I think as soon as the backend PR is merged (no need to wait for deploy), @tienifr can update the PR to use the decimals in Onyx. I'm waiting because I don't want to ask @tienifr to update the PR to use this and then, if it doesn't get approved for some reason, @tienifr has to rollback.

**Comment by aldo-expensify** [comment]
@tienifr the backend PRs has been merged, but not deployed yet. Can you update your PR so it gets the decimal numbers from `currencyList` in Onyx? 🙏 

![image](https://user-images.githubusercontent.com/87341702/231592627-bbc2de22-bea4-4f62-8ab5-91b45c1c0f4e.png)

I'll remove the HOLD once the backend PR is deployed.

**Comment by tienifr** [comment]
@aldo-expensify Pushed the latest update! https://github.com/Expensify/App/pull/16936

**Comment by aldo-expensify** [comment]
PR https://github.com/Expensify/App/pull/16936 merged

**Comment by tienifr** [comment]
# Regression Test Proposal

**Bug**: Split - 'Split bill' incorrectly divides bills for some currencies

**Proposed Test Steps**:

1. Login with any account
2. Click the FAB button and select Split Bill
3. Select currency with no minor unit, i.e. does not have decimal places (for example, AFN or JPY)
4. Enter an amount such that when divided by the number of attendees, fractional values are obtained (i.e. 103 JPY for 3 attendees)
5. Select attendees (i.e. 2 more attendees)
6. Verify that the sum of the splits equals the total amount (i.e. 35 + 34 + 34 = 103)
7. Repeat from step 2
8. Try with 10 JPY and 4 attendees (including your account)
9. Verify that the sum of the splits equals the total amount (i.e. 1 + 3 + 3 + 3 = 10)
10. Now try with currencies having decimals, for example 10 BHD and 3 attendees (including your account)
11. Verify that the sum of the splits equals the total amount (i.e. 3.340 + 3.330 + 3.330 = 10)

![image](https://user-images.githubusercontent.com/113963320/232965616-d5641825-3109-486a-9466-9c8df86fa6d4.png)
![image](https://user-images.githubusercontent.com/113963320/232965814-d8dbb331-3cfc-4cf1-a8c2-2d41d156802e.png)
![image](https://user-images.githubusercontent.com/113963320/232965731-03d110e9-8029-413d-862d-b09c2a23200a.png)

Do we :+1: or :-1:

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.3.2-5 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/16936


If no regressions arise, payment will be issued on 2023-04-27. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@aimane-chnaif] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@aimane-chnaif] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@aimane-chnaif] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@kevinksullivan] Determine if we should create a regression test for this bug.
- [ ] [@aimane-chnaif] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@kevinksullivan] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by kevinksullivan** [comment]
@aimane-chnaif can you please start on the above ^ so we can close this out / make payments on time?

**Comment by kevinksullivan** [comment]
Also sent offers to @aimane-chnaif and @tienifr

**Comment by kevinksullivan** [comment]
All set with payment!
